src/ott/solvers/linear/univariate.py

```diff
@@ -12,134 +12,247 @@
+      return jnp.linspace(0.0, 1.0, self._quantiles)
+    return self._quantiles
+
+  @property
+  def num_quantiles(self):
+    """Number of quantiles used to evaluate OT cost."""
+    return 0 if self._quantiles is None else self.quantiles.shape[0]
 
   def __call__(
       self,
-      x: jnp.ndarray,
-      y: jnp.ndarray,
-      a: Optional[jnp.ndarray] = None,
-      b: Optional[jnp.ndarray] = None
+      prob: linear_problem.LinearProblem,
+      rng: Optional[jax.Array] = None,
   ) -> float:
-    """Computes the Univariate OT Distance between `x` and `y`.
+    """Computes Univariate Distance between the `d` dimensional slices.
 
     Args:
-      x: The first distribution of shape ``[n,]`` or ``[n, 1]``.
-      y: The second distribution of shape ``[m,]`` or ``[m, 1]``.
-      a: The first marginals when ``method = 'wasserstein'``. If :obj:`None`,
-        uniform will be used.
-      b: The second marginals when ``method = 'wasserstein'``. If :obj:`None`,
-        uniform will be used.
+      prob: describing, in its geometry attribute, the two point clouds
+        ``x`` and ``y`` (of respective sizes ``[n,d]`` and ``[m,d]``) and
+        a ground ``cost_fn`` for between two scalars. The ``[n,]`` and ``[m,]``
+        size probability weights vectors are stored in attributes ``a`` and
+        ``b``.
+      rng: used for random downsampling, if used.
+      return_transport: whether to return an average transport matrix (across
+        slices). Not available when approximating the distance computation
+        using subsamples, or quantiles.
 
     Returns:
-      The OT distance.
+      The OT distance, and possibly the transport matrix averaged by
+        considering all matrices arising from 1D transport on each of the ``d``
+        dimensional slices of the input.
     """
-    x = x.squeeze(-1) if x.ndim == 2 else x
-    y = y.squeeze(-1) if y.ndim == 2 else y
-    assert x.ndim == 1, x.ndim
-    assert y.ndim == 1, y.ndim
-
-    n, m = x.shape[0], y.shape[0]
-
-    if self.method == "equal":
-      xx, yy = self.sort_fn(x), self.sort_fn(y)
-    elif self.method == "subsample":
-      assert self.n_subsamples <= n, (self.n_subsamples, x)
-      assert self.n_subsamples <= m, (self.n_subsamples, y)
-
-      sorted_x, sorted_y = self.sort_fn(x), self.sort_fn(y)
-      xx = sorted_x[jnp.linspace(0, n, num=self.n_subsamples).astype(int)]
-      yy = sorted_y[jnp.linspace(0, m, num=self.n_subsamples).astype(int)]
-    elif self.method == "quantile":
-      sorted_x, sorted_y = self.sort_fn(x), self.sort_fn(y)
-      xx = jnp.quantile(sorted_x, q=jnp.linspace(0, 1, self.n_subsamples))
-      yy = jnp.quantile(sorted_y, q=jnp.linspace(0, 1, self.n_subsamples))
-    elif self.method == "wasserstein":
-      return self._cdf_distance(x, y, a, b)
+    geom = prob.geom
+    n, m = geom.shape
+    rng = jax.random.PRNGKey(0) if rng is None else rng
+    geom_is_pc = isinstance(geom, pointcloud.PointCloud)
```

Comment:
Personally I don't like assigning this to a variable, would remove (also in the 2nd instance.

Reply:
sure, wanted to avoid \ :)